gateway: introduce content-security-policy for all responses

Khoyo opened the pull request:
Closes #14523

Khoyo committed:
gateway: introduce content-security-policy for all responses

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

Khoyo committed:
gateway: introduce content-security-policy

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

Khoyo committed:
front: move nge iframe source to a standalone url

This is in order to serve a different CSP for it.

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

Khoyo committed:
front: move nge iframe source to a standalone url

This is in order to serve a different CSP for it.

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

Khoyo committed:
gateway: split up the csp lines

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

emersion commented on front/src/applications/operationalStudies/views/Scenario/components/NGE/NGE.tsx:
```diff
@@ -1,12 +1,5 @@
 import { useEffect, useRef, useState } from 'react';
 
-/* eslint-disable import/extensions */
-import ngeMain from '@osrd-project/netzgrafik-frontend/dist/netzgrafik-frontend/main.js?url';
-import ngePolyfills from '@osrd-project/netzgrafik-frontend/dist/netzgrafik-frontend/polyfills.js?url';
-import ngeRuntime from '@osrd-project/netzgrafik-frontend/dist/netzgrafik-frontend/runtime.js?url';
-import ngeStyles from '@osrd-project/netzgrafik-frontend/dist/netzgrafik-frontend/styles.css?url';
-import ngeVendor from '@osrd-project/netzgrafik-frontend/dist/netzgrafik-frontend/vendor.js?url';
-/* eslint-enable import/extensions */
```
I think this might break hot reload with NGE as an `npm link` package?

Khoyo replied:
Would keeping the import but not using them be better?

emersion replied:
This has been dropped in cb07c7c5eb84cd6db259563da8d69f8e0c0bd985 anyways, so this thread is moot.

emersion commented on gateway/src/csp.rs:
```diff
@@ -0,0 +1,83 @@
+    forward_ready!(service);
+
+    fn call(&self, req: ServiceRequest) -> Self::Future {
+        let csp_string = if req.path().starts_with("/netzgrafik-frontend") {
```
I think we can narrow this down to `/netzgrafik-frontend/` (with a final slash), just in case.

emersion commented on gateway/src/csp.rs:
```diff
@@ -0,0 +1,83 @@
+            // FIXME: add unsafe-inline for scripts if we don't keep the next commit
+            // or, ideally, fix NGE to not require it.
+            concat!(
+                "default-src 'self'; form-action 'self'; frame-ancestors 'self'; ",
```
Do we need to explicitly specify everything that's identical to `default-src`? Wouldn't these already be set to `'self'` by default?

Khoyo replied:
Nope. `default-src` only applies to a [select few directives](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/default-src). See eg. [form-action](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/form-action)

> [default-src](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/default-src) fallback | No. Not setting this allows anything.

(tbf you're right for `style-src`)

emersion replied:
Oh. TIL!

Khoyo committed:
front, gateway: remove the meta_env import

It's not used anymore, and we need to remove inline scripts in
preparation for a strict CSP.

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

emersion commented on front/public/netzgrafik-frontend/index.html:
```diff
@@ -0,0 +1,10 @@
+<!doctype html>
+<html class="sbb-lean sbb-light">
+  <head>
+    <base href="/netzgrafik-frontend/">
```
Now that this lives inside the same directory as other assets, I think we can drop the base and use relative imports for the rest?

emersion commented on front/src/import_meta_env.js:
```diff
@@ -0,0 +1 @@
+globalThis.import_meta_env = JSON.parse('"import_meta_env_placeholder"');
```
Is this used from anywhere? I think it can just be dropped, we've migrated away from import-meta-env and we now use Vite to pass env vars.

Khoyo replied:
Nice
